### ISSUE TITLE
Discussion: checking in vendored code (Node modules, Go vendor)

### ISSUE THREAD

**Issue by radu-matei** [brigadecore/brigade#930]
https://github.com/brigadecore/brigade/commit/ad4152af0cee154057878ed520a9a2e1bc5a60ee checked in the `node_modules` directory.

We should probably remove it from the repository.

**Comment by krancour** [comment]
I strongly disagree with not tracking vendored code whenever possible. It's extremely onerous and easily overlooked for contributors to re-run dependency resolution every time they switch branches. Even if it's automated so that, for instance, dependencies are automatically resolved before a build or test, that can slow those tasks down substantially-- especially in the event that they are occurring within a container.

By contrast, when vendored code is tracked, the moment you clone a repo or checkout a given branch, you're ready to build.

All of this being said, I'm not going to choose this hill to die on. If others are vehemently opposed to this, we can undo it. I would, however, suggest that if this is somehow causing issues for people that we try to address those issues rather than jump right to saying that we _shouldn't_ track vendored code.

**Comment by radu-matei** [comment]
One point to counter this is: if you clone this repository on a different operating system and try to run something, the binaries are built for a different architecture, so it's not going to work. You `yarn install` again, but now you introduced changes that are going to break CI.

**Comment by itowlson** [comment]
@krancour I'm not a Brigade maintainer and it's up to you folks to make a decision on this, but Radu asked me to comment on this as a disinterested observer.

As a Node user, I would be quite surprised to clone a repo and find myself pulling down the whole `node_modules` directory.  I'm used to running `npm install` every time I change branches (and yes, it is annoying!).  The Node idiom seems to be not to check in the modules, as witness the fact that generated templates (`npm init`, GitHub, `yo code`) have `node_modules` in their .gitignore.  This of course is an argument from habit, not on the technical merits, though.

Some packages have install actions which would be bypassed if the package were simply cloned.  For example, the `vscode` package runs a binary at install time to pull down the TypeScript definitions appropriate to the VS Code engine in use.

Including vendored dependencies can also lead to large PRs, where an updated or new dependency (and its transitive dependencies) can end up swamping the actual application changes, making the PR hard to review.  This is arguably an issue with PR UIs, but it's still something to reckon with.

Finally some enterprises will have policies to load packages from trusted internal feeds (particularly during official builds in compliance environments).  Cloning packages from the Git repo bypasses those policies.  It may even be worse - suppose a bad actor sent a legitimate PR that introduced a legitimate dependency on a trusted package, but the bad actor went into `node_modules` and tampered with the package or a dependency there (NOT hijacking the package on NPM - just editing their local files).  If that got checked in, everyone who pulled the commit would get the maliciously tampered package.  So now you not only have to review 'do the dependencies look legit' but 'read all the code in all the dependencies and confirm that is what is in NPM'.

On the plus side, checking in modules would avoid possible issues with changes.  These could be somewhat mitigated by locking versions in `package-lock.json`, but you still have the left-pad scenario where a package goes away - having your own copies of dependencies ensures that will never happen.

Just some thoughts - hope this helps.

**Comment by krancour** [comment]
@itowlson I should lead by saying that this isn't a hill I'm going to die on. You make some excellent points. I do have some counter-points. In the end, I don't care what our collective answer to this particular conundrum is. What I care about the the _overall_ experience for contributors. As long as that is adequately hassle-free, I am happy-- regardless of what shape it takes.

> As a Node user, I would be quite surprised to clone a repo and find myself pulling down the whole node_modules directory. I'm used to running npm install every time I change branches (and yes, it is annoying!). The Node idiom seems to be not to check in the modules, as witness the fact that generated templates (npm init, GitHub, yo code) have node_modules in their .gitignore. This of course is an argument from habit, not on the technical merits, though.

I am not a frequent-enough or proficient-enough user of Node to speak to what is typical within that community. I defer to your observation of trends here. That being said, JS / Node code accounts for a slim minority of the code in this repository, with Go accounting for the rest. So the question of what is familiar to Go developers may (for better or worse) also be pertinent since such developers are likely to account for many of the contributors. The Go community also argues over whether dependencies should be vendored or not. My finger is on the pulse of that community more so than the Node community and my current read on things is that many agree that dependencies should not be vendored/tracked for _libraries_, but should be for binaries/applications, etc. Put a pin in that for a moment...

> Some packages have install actions which would be bypassed if the package were simply cloned. For example, the vscode package runs a binary at install time to pull down the TypeScript definitions appropriate to the VS Code engine in use.

An excellent point, but I believe this is probably rendered moot by the fact that all development-related tasks (e.g. build, test, and even dependency resolution) aren't executing on the contributor's host OS, but rather in a containerized development sandbox. If post-install actions invoke (or even build) binaries, we're invoking/building binaries appropriate for linux/64, which is also the OS/arch that we target deployment to (i.e. that's the OS/arch of the Brigade worker Docker image). Assuming post-install actions limit their changes to the `node_modules/` tree, there's no problem here, as the results of the post-install action will, themselves, be vendored/tracked. If that isn't the case and those post-install actions make _global_ system modifications to the (ephemeral) container, then yes, there is a problem... but builds that are dependent on system state have a major problem with portability and repeatability to begin with, so I'll assume post-install actions that make global system changes are both bad and rare.

> Including vendored dependencies can also lead to large PRs, where an updated or new dependency (and its transitive dependencies) can end up swamping the actual application changes, making the PR hard to review. This is arguably an issue with PR UIs, but it's still something to reckon with.

I cannot argue with this. You're right. fwiw, this can be mitigated by isolating changes to vendored/tracked dependencies to their own commit(s), which can be effectively disregarded by reviewers who can then focus the remaining commits that contain reviewable changes. Granted, there is no way to enforce people adopting this approach. At least, in my experience, sweeping updates to dependencies aren't common by contributors other than the maintainers. 🤷‍♂ 

> Finally some enterprises will have policies to load packages from trusted internal feeds (particularly during official builds in compliance environments). Cloning packages from the Git repo bypasses those policies.

I've worked in such an enterprise. Whether that particular policy is bypassed by our approach is debatable. Won't they, after all, have obtained our own code from the sanctioned source? I realize there's not an easy yes/no answer to this. It will depend on the _exact_ policy is and how strict it is. That being said... I return to the distinction I made earlier between library and application/binary. All of our Node code is shipped as a Docker image. I would think, in the end, few users who are not contributors will ever entertain building the image from source, rather, they will use pre-built Docker images. That being the case, enterprise policies concerning the source/pedigree/integrity of Docker images would likely trump any policy concerning the source/pedigree/integrity of the libraries utilized _by_ that image. Again, this is all hypothetical, and depends a lot on the org and their exact policies. imho, it's not possible to constrain ourselves such that we satisfy the infinite range of hypothetical, uber-restrictive enterprise policies.

> suppose a bad actor sent a legitimate PR that introduced a legitimate dependency on a trusted package, but the bad actor went into node_modules and tampered with the package or a dependency there (NOT hijacking the package on NPM - just editing their local files). If that got checked in, everyone who pulled the commit would get the maliciously tampered package. So now you not only have to review 'do the dependencies look legit' but 'read all the code in all the dependencies and confirm that is what is in NPM'.

This is a valid concern that is mitigated by an automated check on the integrity of the vendored code. i.e. Does the computed sha of the package match what it's supposed to be. This is already in place.

In summary, building, in particular, off the second-to-last point above about how this is generally not built from source by _users_, I return to the only strong assertion I made, which is that my ultimate concern is for optimizing clone, build, test, dependency resolution, etc. for _contributors_.

I want to leave the ultimate decision here to @radu-matei since I think he's probably most invested in this project, but my humble suggestion would be that we leave this be for the near term in order to gain a real-world perspective on how this is positively or negatively impacting the workflow for maintainers and other contributors.

**Comment by radu-matei** [comment]
It has been almost one year since we initially started this conversation, and I think now it be a great time to re-open the discussion.

#1075 added support for Go modules, and I am not sure I follow still having a `vendor` directory.
Additionally, the points `node_modules` not usually being checked-in in the Node ecosystem are still valid - and so now we are in a situation where both a Go and Node developer would be surprised to see the vendored code directories in the repository.

However, @krancour's points about the ease of having the directories when building everything in a containerized environment are also extremely on point and valid.

And so I believe the main question is - are the advantages gained by running in a container important enough to justify going against "established" practices in both Go and Node communities, and the friction that comes with building in a container (time and resources)?

Personally, I think not, and I would advocate for building on the host OS _by default_, but I would also like to understand what other maintainers and contributors think, and understand if this approach has improved the developer experience.

**Comment by krancour** [comment]
@radu-matei aren't the questions of whether we build in containers and whether we vendor source code actually two different questions? Related in some ways? Of course. But I'm not sure the answer to one must dictate the answer to the other. Thoughts?

Also, I think, at least where Go is concerned, the recent switch to using go mod may help us here because, I believe, all other go commands will automatically carry out dependency resolution as needed as a prereq. That does a lot to alleviate my concerns about, for instance, changing branches and accidentally shooting yourself in the foot by forgetting to explicitly invoke some dependency resolution step.

fwiw... I can much more easily be convinced not to track vendored code than to abandon containerized builds/tests etc. The builds run in containers in CI. Given the opportunity, they should run in the same container locally for maximum parity.

**Comment by krancour** [comment]
We haven't been vendoring node dependencies since #1431.

v1.x still vendors Go dependencies (although v2 does not).

Going to close this on account of staleness, but please do reopen if it warrants revisiting.
